File: packages/types/src/usage.ts
/**
 * Usage reporting types for provider quota/limit endpoints.
 * Ported from oh-my-pi/packages/ai/src/usage.ts — trimmed for console.
 */
import type { ProviderId } from "./model";

export type UsageUnit = "percent" | "tokens" | "requests" | "usd" | "minutes" | "bytes" | "unknown";
export type UsageStatus = "ok" | "warning" | "exhausted" | "unknown";

export interface UsageWindow {
  id: string;
  label: string;
  durationMs?: number;
  resetsAt?: number;
  resetLabel?: string;
}

export interface UsageAmount {
  used?: number;
  limit?: number;
  remaining?: number;
  usedFraction?: number;
  remainingFraction?: number;
  unit: UsageUnit;
}

export interface UsageScope {
  provider: ProviderId;
  accountId?: string;
  projectId?: string;
  orgId?: string;
  modelId?: string;
  tier?: string;
  windowId?: string;
  shared?: boolean;
}

export interface UsageLimit {
  id: string;
  label: string;
  scope: UsageScope;
  window?: UsageWindow;
  amount: UsageAmount;
  status?: UsageStatus;
  notes?: string[];
}

export interface UsageResetCreditDetail {
  grantedAt?: string;
  expiresAt?: string;
  status?: string;
}

export interface UsageResetCredits {
  availableCount: number;
  credits?: UsageResetCreditDetail[];
}

export interface UsageReport {
  provider: ProviderId;
  fetchedAt: number;
  limits: UsageLimit[];
  resetCredits?: UsageResetCredits;
  notes?: string[];
  metadata?: Record<string, unknown>;
  raw?: unknown;
}

export function resolveUsedFraction(limit: UsageLimit): number | undefined {
  const amount = limit.amount;
  if (amount.usedFraction !== undefined) return amount.usedFraction;
  if (amount.used !== undefined && amount.limit !== undefined && amount.limit > 0) {
    return amount.used / amount.limit;
  }
  if (amount.unit === "percent" && amount.used !== undefined) return amount.used / 100;
  if (amount.remainingFraction !== undefined) return Math.max(0, 1 - amount.remainingFraction);
  return undefined;
}

export interface UsageLogger {
  debug(message: string, meta?: Record<string, unknown>): void;
  warn(message: string, meta?: Record<string, unknown>): void;
}

export interface UsageCredential {
  type: "api_key" | "oauth";
  apiKey?: string;
  accessToken?: string;
  refreshToken?: string;
  expiresAt?: number;
  accountId?: string;
  projectId?: string;
  email?: string;
  orgId?: string;
  orgName?: string;
  enterpriseUrl?: string;
  metadata?: Record<string, unknown>;
  apiEndpoint?: string;
}

export interface UsageFetchParams {
  provider: ProviderId;
  credential: UsageCredential;
  accountKey?: string;
  baseUrl?: string;
  signal?: AbortSignal;
}

export type FetchImpl = typeof fetch;

export interface UsageFetchContext {
  fetch: FetchImpl;
  logger?: UsageLogger;
  retryWait?: (delayMs: number, signal?: AbortSignal) => Promise<void>;
}

export interface UsageProvider {
  id: ProviderId;
  fetchUsage(params: UsageFetchParams, ctx: UsageFetchContext): Promise<UsageReport | null>;
  parseRateLimitHeaders?(headers: Record<string, string>, now?: number): UsageReport | null;
  supports?(params: UsageFetchParams): boolean;
  validatesCredentials?: boolean;
  retainLastGoodOnFailure?: boolean;
}

export interface CredentialRankingContext {
  modelId?: string;
}

export interface CredentialRankingStrategy {
  findWindowLimits(
    report: UsageReport,
    context?: CredentialRankingContext,
  ): {
    primary?: UsageLimit;
    secondary?: UsageLimit;
  };
  scopeLimits?(report: UsageReport, context?: CredentialRankingContext): UsageLimit[];
  scopeLimitsForReserve?(report: UsageReport, context?: CredentialRankingContext): UsageLimit[];
  blockScope?(context?: CredentialRankingContext): string | undefined;
  blockScopes?(context?: CredentialRankingContext): string[];
  windowDefaults: {
    primaryMs: number;
    secondaryMs: number;
  };
  hasPriorityBoost?(primary: UsageLimit | undefined): boolean;
}

File: apps/desktop/crates/console-core/src/types/usage.rs
//! Usage and quota DTOs mirroring server `/api/usage` and `@console/types`.

use serde::{Deserialize, Serialize};

#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum UsageUnit {
    Percent,
    Tokens,
    Requests,
    Usd,
    Minutes,
    Bytes,
    #[serde(other)]
    Unknown,
}

impl Default for UsageUnit {
    fn default() -> Self {
        Self::Unknown
    }
}

#[derive(Clone, Copy, Debug, Eq, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum UsageStatus {
    Ok,
    Warning,
    Exhausted,
    #[serde(other)]
    Unknown,
}

impl Default for UsageStatus {
    fn default() -> Self {
        Self::Unknown
    }
}

#[derive(Clone, Debug, Default, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct UsageWindow {
    pub id: String,
    pub label: String,
    #[serde(default)]
    pub duration_ms: Option<i64>,
    #[serde(default)]
    pub resets_at: Option<i64>,
    #[serde(default)]
    pub reset_label: Option<String>,
}

#[derive(Clone, Debug, Default, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct UsageAmount {
    #[serde(default)]
    pub used: Option<f64>,
    #[serde(default)]
    pub limit: Option<f64>,
    #[serde(default)]
    pub remaining: Option<f64>,
    #[serde(default)]
    pub used_fraction: Option<f64>,
    #[serde(default)]
    pub remaining_fraction: Option<f64>,
    #[serde(default)]
    pub unit: UsageUnit,
}

#[derive(Clone, Debug, Default, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct UsageScope {
    pub provider: String,
    #[serde(default)]
    pub account_id: Option<String>,
    #[serde(default)]
    pub project_id: Option<String>,
    #[serde(default)]
    pub org_id: Option<String>,
    #[serde(default)]
    pub model_id: Option<String>,
    #[serde(default)]
    pub tier: Option<String>,
    #[serde(default)]
    pub window_id: Option<String>,
    #[serde(default)]
    pub shared: Option<bool>,
}

#[derive(Clone, Debug, Default, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct UsageLimit {
    pub id: String,
    pub label: String,
    pub scope: UsageScope,
    #[serde(default)]
    pub window: Option<UsageWindow>,
    pub amount: UsageAmount,
    #[serde(default)]
    pub status: Option<UsageStatus>,
    #[serde(default)]
    pub notes: Option<Vec<String>>,
}

impl UsageLimit {
    pub fn resolved_used_fraction(&self) -> f64 {
        if let Some(frac) = self.amount.used_fraction {
            return frac.clamp(0.0, 1.0);
        }
        if let (Some(used), Some(limit)) = (self.amount.used, self.amount.limit) {
            if limit > 0.0 {
                return (used / limit).clamp(0.0, 1.0);
            }
        }
        if self.amount.unit == UsageUnit::Percent {
            if let Some(used) = self.amount.used {
                return (used / 100.0).clamp(0.0, 1.0);
            }
        }
        if let Some(rem_frac) = self.amount.remaining_fraction {
            return (1.0 - rem_frac).clamp(0.0, 1.0);
        }
        0.0
    }

    pub fn resolved_remaining_fraction(&self) -> f64 {
        if let Some(rem) = self.amount.remaining_fraction {
            return rem.clamp(0.0, 1.0);
        }
        1.0 - self.resolved_used_fraction()
    }
}

#[derive(Clone, Debug, Default, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct UsageReport {
    pub provider: String,
    pub fetched_at: i64,
    #[serde(default)]
    pub limits: Vec<UsageLimit>,
    #[serde(default)]
    pub notes: Option<Vec<String>>,
    #[serde(default)]
    pub metadata: Option<serde_json::Value>,
}
